include/bluetooth/services/ras.h

```diff
@@ -130,6 +131,82 @@ struct ras_subevent_header {
 } __packed;
 BUILD_ASSERT(sizeof(struct ras_subevent_header) == BT_RAS_SUBEVENT_HEADER_LEN);
 
+/** @brief Ranging data ready callback. Called when peer has ranging data available.
+ *
+ * @param[in] conn            Connection Object.
+ * @param[in] ranging_counter Ranging counter ready to be requested.
+ */
+typedef void (*bt_ras_rreq_rd_ready_cb_t)(struct bt_conn *conn, uint16_t ranging_counter);
+
+/** @brief Ranging data overwritted callback. Called when peer has overwritten previously available
+ * ranging data.
+ *
+ * @param[in] conn            Connection Object.
+ * @param[in] ranging_counter Ranging counter which has been overwritten.
+ */
+typedef void (*bt_ras_rreq_rd_overwritten_cb_t)(struct bt_conn *conn, uint16_t ranging_counter);
+
+/** @brief Ranging data get complete callback. Called when ranging data get procedure has completed.
+ *
+ * @param[in] conn            Connection Object.
+ * @param[in] ranging_counter Ranging counter which has been overwritten.
+ * @param[in] err             Error code, 0 if the ranging data get was successful. Otherwise a
+ * negative error code.
+ */
+typedef void (*bt_ras_rreq_ranging_data_get_complete_t)(struct bt_conn *conn,
+							uint16_t ranging_counter, int err);
+
+/** @brief Allocate a RREQ context and assign GATT handles. Takes a reference to the connection.
+ *
+ * @param[in] dm   Discovery Object.
+ * @param[in] conn Connection Object.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a negative error code is returned.
+ */
+int bt_ras_rreq_alloc_and_assign_handles(struct bt_gatt_dm *dm, struct bt_conn *conn);
+
+/** @brief Get ranging data for given ranging counter.
+ *
+ * @note This should only be called after receiving a ranging data ready callback and
+ * when subscribed to ondemand ranging data and RAS-CP.
+ *
+ * @param[in] conn                 Connection Object.
+ * @param[in] ranging_data_out     Simple buffer to store received ranging data.
+ * @param[in] ranging_counter      Ranging counter to get.
+ * @param[in] data_get_complete_cb Callback called when get ranging data completes.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a negative error code is returned.
+ */
+int bt_ras_rreq_cp_get_ranging_data(struct bt_conn *conn, struct net_buf_simple *ranging_data_out,
+				    uint16_t ranging_counter,
+				    bt_ras_rreq_ranging_data_get_complete_t data_get_complete_cb);
+
+/** @brief Free RREQ context for connection. Should be called from disconnected callback. This will
```

Comment:
is it possible to free it automatically when the link disconnects ? Otherwise I think I would forget

Reply:
I could align similar to how Alex has done it with CONFIG_BT_RAS_RRSP_AUTO_ALLOC_INSTANCE?

Reply:
would it work to just call `bt_ras_rreq_free()` in the disconnected callback ? Then if no RREQ context was allocated by the app it just returns here with no effect https://github.com/nrfconnect/sdk-nrf/pull/18737/files/3c6272b8be2d982db4a094367f742de0fc1f15f2#diff-3f67cd569279ac64cbc075871eff1096b3735294104897f17ca416a75866ace7R563 ?

Reply:
Yeah good point